WIP: ENH: ns datetime "now"

tylerjereddy opened the pull request:
Fixes #13372

* use `_PyTime_GetSystemClock()`, which provides nanosecond
resolution time in a portable manner via the Python C
API, to provide high resolution time information for
`np.datetime64("now", "ns")`

* note that the displays differ slightly here
in the number of digits after the decimal (I
believe the 000 convention is normal for datetime
reporting, but we may want to adjust that, not sure):
```
np.datetime64(datetime.utcnow())
np.datetime64("now", "ns")
2021-07-24T23:52:33.303293
2021-07-24T23:52:33.303595000
```

* we may want to consider using `_PyTime_GetSystemClockWithInfo()`,
which actually returns a `-1` we could use as a handle for
some degree of error checking

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

tylerjereddy committed:
ENH: ns datetime "now"

* use `_PyTime_GetSystemClock()`, which provides nanosecond
resolution time in a portable manner via the Python C
API, to provide high resolution time information for
`np.datetime64("now", "ns")`

* note that the displays differ slightly here
in the number of digits after the decimal (I
believe the 000 convention is normal for datetime
reporting, but we may want to adjust that, not sure):
```
np.datetime64(datetime.utcnow())
np.datetime64("now", "ns")
2021-07-24T23:52:33.303293
2021-07-24T23:52:33.303595000
```

* we may want to consider using `_PyTime_GetSystemClockWithInfo()`,
which actually returns a `-1` we could use as a handle for
some degree of error checking

tylerjereddy commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -323,21 +324,20 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
         PyArray_DatetimeMetaData meta;
 
-        time(&rawtime);
+        int64_t rawtime = _PyTime_GetSystemClock();
```
some discussion in [PEP564](https://www.python.org/dev/peps/pep-0564/)

tylerjereddy commented on numpy/core/tests/test_datetime.py:
```diff
@@ -2446,6 +2446,15 @@ def test_limit_str_roundtrip(self, time_unit, sign):
+        # before supporting ns resolution, the
+        # second-truncated newer_stamp would
+        # fail this test
+        assert_array_less(stamp, newer_stamp)
```
This regression test always failed on `main` for me locally, though it does occur to me that if you happen to run the test at x.99999 time units the clock might switch over and pass the test accidentally. I suspect the likelihood of that happening is on lottery ticket level probability, but haven't thought too hard about it.

There's probably a more obvious hook I could use to check--the repr not having all zeros at the end for example (although, you could accidentally have that pass as well if you just happen to hit 00000 ns or whatever).

mattip commented:
Apparently this needs some tweaking for python3.10? The [build fails](https://github.com/numpy/numpy/pull/19558/checks?check_run_id=3152669479#step:4:1248) with 
```
numpy/core/src/multiarray/datetime_strings.c:14:10: fatal error: pytime.h: No such file or directory
   14 | #include <pytime.h>
      |          ^~~~~~~~~~
compilation terminated.
```

The [release notes](https://docs.python.org/3.10/whatsnew/3.10.html) mention
> The non-limited API files ..., and pytime.h ... must not be included directly, as they are already included in Python.h ..., [C]onsider including Python.h instead.

eric-wieser commented:
I think it would be cool if `datetime64("now")` used the `resolution` result from `_PyTime_GetSystemClockWithInfo` to pick an appropriate precision.

charris commented:
Not sure what the error is on 3.10. @tylerjereddy could you try a rebase to see if that fixes things?

tylerjereddy commented:
I think Matti found the 3.10 problem above; in the raw log it says `numpy/core/src/multiarray/datetime_strings.c:14:10: fatal error: pytime.h: No such file or directory` buried deep in the build output.

Presumably I can fix that easily enough--I think there are numerous examples of NumPy C code making decisions based on Python version.

I was going to try doing what Eric suggests as well, though that may take a bit more thought on user specified vs. max available precision handling (maybe not too bad, not sure).

tylerjereddy committed:
MAINT: PR 19558 revisions

* guard the `pytime.h` include based on
Python version

* early draft of using time resolution information
from `_PyTime_GetSystemClockWithInfo()` to make
better decisions with `datetime64("now")`

tylerjereddy commented:
I tried to move this a little closer to the reviewer suggestions, but certainly needs more work. I'm pulling out resolution information now, but it feels pretty awkward at the moment. Some sample behavior below--it looks like the resolution I get locally with the C function is about `1e-6` (perf counter is higher when I check it, but maybe that makes sense for a draw from the clock vs. perf counter?).

```python
import time
import numpy as np

print("np.__version__:", np.__version__)

for unit in ['s', 'ms', 'us', '1000ns', '10ns', '100ns', 'ns']:
    print(np.datetime64("now", unit), f"(unit requested: {unit})")

print(np.datetime64("now"), f"(unit requested: None)")

std_result = time.get_clock_info('time')
print("std resolution:", std_result.resolution)

perf_result = time.get_clock_info('perf_counter')
print("perf resolution:", perf_result.resolution)
```

```
np.__version__: 1.22.0.dev0+567.g9de031dce
2021-08-08T22:23:56 (unit requested: s)
2021-08-08T22:23:56.925 (unit requested: ms)
2021-08-08T22:23:56.925496 (unit requested: us)
2021-08-08T22:23:56.925500000 (unit requested: 1000ns)
2021-08-08T22:23:56.925503000 (unit requested: 10ns)
2021-08-08T22:23:56.925506000 (unit requested: 100ns)
2021-08-08T22:23:56.925509000 (unit requested: ns)
2021-08-08T22:23:56.925513000 (unit requested: None)
std resolution: 1.0000000000000002e-06
perf resolution: 1e-09
```

tylerjereddy commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -323,22 +326,37 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
+        }
+        else if (resolution < 0.0000000011) {
+            bestunit = NPY_FR_ps;
+        }
```
This block feels like a terrible hack--I'm not sure what span of resolution units we reasonably need to check for, but surely there is a cleaner way to do the float comparisons, etc.

tylerjereddy commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -323,22 +326,37 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
+        // actual rawtime we use is from here,
+        // along with resolution info
+        _PyTime_GetSystemClockWithInfo(&rawtime, &rawtime_info);
+        double resolution = rawtime_info.resolution;
```
`1e-6` when I happened to `printf` it locally

tylerjereddy commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -323,22 +326,37 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
 
-        time(&rawtime);
+        // initialize the _PyTime_t struct
+        _PyTime_t rawtime = _PyTime_GetSystemClock();
```
I was getting segfaults when I set this just to some arbitrary positive `int64_t` value like `0` or `1` I think. Maybe setting it to some reasonably large integer offset from the epoch might be safe and avoid two time draws in the control flow?

tylerjereddy commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -323,22 +326,37 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
 
         /* Set up a dummy metadata for the conversion */
-        meta.base = NPY_FR_s;
+        meta.base = bestunit;
```
I think I need a better understanding of the control flow for "generic case--user doesn't specify a time unit so make the best choice" vs. "use the unit the user specified."

Maybe I'm getting a little closer...

tylerjereddy commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -323,22 +326,37 @@ parse_iso_8601_datetime(char const *str, Py_ssize_t len,
+        else if ((resolution < 0.001) && (resolution > 0.0000009)) {
+            bestunit = NPY_FR_ns;
+        }
+        else if (resolution < 0.0000000011) {
```
not using a plain `else` here is setting off the CI smoke test, but it is early draft anyway

tylerjereddy commented:
Let me close this in case someone else wants to drive it forward. I may focus on something a bit smaller in NumPy for now.
